sql/parser/eval.go

```diff
@@ -319,109 +319,236 @@ type cmpArgs struct {
-		pattern = strings.Replace(pattern, "_", ".", -1)
-		pattern = anchorPattern(pattern, false)
+	cmpArgs{Like, stringType, stringType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			re := cache.(*regexp.Regexp)
+			return DBool(re.MatchString(string(left.(DString)))), nil
+		},
+		initFn: func(left Datum, right Datum) (cmpOpCacheVal, error) {
+			pattern := regexp.QuoteMeta(string(right.(DString)))
+			// Replace LIKE specific wildcards with standard wildcards
+			pattern = strings.Replace(pattern, "%", ".*", -1)
+			pattern = strings.Replace(pattern, "_", ".", -1)
+			pattern = anchorPattern(pattern, false)
+
+			return regexp.MustCompile(pattern), nil
+		},
+	},
 
-		// TODO(pmattis): Can we cache this regexp compilation?
-		re := regexp.MustCompile(pattern)
-		return DBool(re.MatchString(string(left.(DString)))), nil
+	cmpArgs{SimilarTo, stringType, stringType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			re := cache.(*regexp.Regexp)
+			return DBool(re.MatchString(string(left.(DString)))), nil
+		},
+		initFn: func(left Datum, right Datum) (cmpOpCacheVal, error) {
+			pattern := SimilarEscape(string(right.(DString)))
+			pattern = anchorPattern(pattern, false)
+
+			return regexp.Compile(pattern)
+		},
 	},
+}
 
-	cmpArgs{SimilarTo, stringType, stringType}: func(left Datum, right Datum) (DBool, error) {
-		pattern := SimilarEscape(string(right.(DString)))
-		pattern = anchorPattern(pattern, false)
+var evalTupleEQ = cmpOp{
+	fn: func(ldatum, rdatum Datum, cache cmpOpCacheVal) (DBool, error) {
+		left := ldatum.(DTuple)
+		right := rdatum.(DTuple)
+		if len(left) != len(right) {
+			return DBool(false), nil
+		}
+		for i := range left {
+			d, err := evalComparisonEq(left[i], right[i])
+			if err != nil {
+				return DummyBool, err
+			}
+			if v, err := getBool(d); err != nil {
+				return DummyBool, err
+			} else if !v {
+				return v, nil
+			}
+		}
+		return DBool(true), nil
+	},
+}
 
-		// TODO(pmattis): Can we cache this regexp compilation?
-		re, err := regexp.Compile(pattern)
-		if err != nil {
-			return DBool(false), err
+var evalTupleIN = cmpOp{
+	fn: func(arg, values Datum, cache cmpOpCacheVal) (DBool, error) {
+		if arg == DNull {
+			return DBool(false), nil
+		}
+
+		vtuple := values.(DTuple)
+
+		// TODO(pmattis): If we're evaluating the expression multiple times we should
+		// use a map when possible. This works as long as arg is not a tuple. Note
+		// that the usage of the map is currently disabled via the "&& false" because
+		// building the map is a pessimization if we're only evaluating the
+		// expression once. We need to determine when the expression will be
+		// evaluated multiple times before enabling. Also need to figure out a way to
+		// use the map approach for tuples. One idea is to encode the tuples into
+		// strings and then use a map of strings.
+		if _, ok := arg.(DTuple); !ok && false {
```

Comment:
I know this is just code movement for this change, but it would be nice to address this TODO in a subsequent PR. It was written a while ago and is a bit out of date. `NormalizeExpr` now guarantees that the values in a tuple or sorted and we should take advantage of that here and perform a binary search. There is code in `normalize.go` which demonstrates how to do this and could be refactored for use here. Once we can perform a binary search for `IN` expressions I think the map construction code can be removed.